typescript/src/electrum.ts

```diff
@@ -120,19 +138,50 @@ export class Client implements BitcoinClient {
    * @returns Promise holding the outcome.
    */
   private async withElectrum<T>(action: Action<T>): Promise<T> {
-    const electrum = new Electrum(
-      this.credentials.host,
-      this.credentials.port,
-      this.credentials.protocol,
-      this.options
-    )
+    const connect = async (credentials: Credentials): Promise<Electrum> => {
+      const electrum: Electrum = new Electrum(
+        credentials.host,
+        credentials.port,
+        credentials.protocol,
+        this.options
+      )
 
-    try {
       await this.withBackoffRetrier()(async () => {
-        return await electrum.connect("tbtc-v2", "1.4.2")
+        // FIXME: Connection timeout should be a property of the Electrum client.
+        // Since it's not configurable in `electrum-client-js` we add timeout
+        // as a workaround here.
+        return pTimeout(
+          (async () => {
+            try {
+              await electrum.connect("tbtc-v2", "1.4.2")
+              await electrum.server_ping()
+              return
+            } catch (error) {
+              throw new Error(`Electrum server connection failure: [${error}]`)
+            }
+          })(),
+          this.connectionTimeout,
+          `timed out on electrum connect after ${this.connectionTimeout} ms`
+        )
       })
-    } catch (error) {
-      throw new Error(`Electrum server connection failure: [${error}]`)
+
+      return electrum
+    }
+
+    let electrum: Electrum | undefined = undefined
+    for (const credentials of this.credentials) {
+      try {
+        electrum = await connect(credentials)
+        break
```

Comment:
Do we always throw an error because of any trouble with the connection? If a connection returns because of some other reason and will not **throw** an error, then I think this `break` would prevent taking another credential from the credentials array.

Reply:
Within `connect` function we try to establish connection to the server and send a `server_ping` request to test it. In case of problems or timeout we throw an error:
https://github.com/keep-network/tbtc-v2/blob/feb7cfe1c98a68a2ec8f19749926e12d9f39ad63/typescript/src/electrum.ts#L149-L168

Reply:
Ok, as long as the error is thrown and not just simply returned, then I think we're fine.